File: scan-library-tool/walker.py
"""Filesystem walker — scans library root for .pdf files only.
   Supports parallel directory scanning via ThreadPoolExecutor.
   Optional XXH3_128 hash cache to skip inline hash computation."""

import json
import os
from collections.abc import Callable
from concurrent.futures import ThreadPoolExecutor, as_completed
from datetime import datetime, timezone
from pathlib import Path

try:
    import xxhash
    _has_xxhash = True
except ImportError:
    _has_xxhash = False

_HASH_CACHE: dict[str, str] | None = None


def load_hash_cache(path: str | None = None) -> int:
    """Load the Rust-generated XXH3_128 hash cache.

    Pass path directly. No-op if file doesn't exist. The env var HASH_CACHE_DIR
    is used only as a fallback when path is None (for CLI usage).
    Returns number of entries loaded (0 if cache not found or corrupted).
    """
    global _HASH_CACHE
    if path is None:
        path = os.environ.get("HASH_CACHE_DIR", "")
    if not path or not os.path.isfile(path):
        return 0
    try:
        with open(path) as f:
            data = json.load(f)
        entries = data.get("entries", {})
        _HASH_CACHE = {uri: e["hash"] for uri, e in entries.items()}
        print(f"[walker] Loaded hash cache: {len(_HASH_CACHE)} entries from {path}")
        return len(_HASH_CACHE)
    except Exception as e:
        print(f"[walker] Failed to load hash cache from {path}: {e}")
        return 0


def _mtime(path: Path) -> str:
    return datetime.fromtimestamp(
        path.stat().st_mtime, tz=timezone.utc
    ).isoformat()


def _hash(path: Path) -> str:
    if _HASH_CACHE is not None:
        uri = path.as_uri()
        if uri in _HASH_CACHE:
            return _HASH_CACHE[uri]
    if not _has_xxhash:
        raise RuntimeError("xxhash package required for file hashing: pip install xxhash")
    h = xxhash.xxh3_128(seed=0)
    with open(path, "rb") as f:
        while chunk := f.read(65536):
            h.update(chunk)
    return h.hexdigest()


def _walk_one_dir(entry: Path, hash_files: bool) -> tuple[str, dict | None]:
    """Process one series directory. Thread-safe — no shared mutable state."""
    try:
        books = []
        for f in sorted(entry.rglob("*"), key=lambda e: e.name.lower()):
            if f.is_file() and f.suffix.lower() == ".pdf":
                book = {
                    "url": f.as_uri(),
                    "name": f.stem,
                    "file_size": f.stat().st_size,
                    "file_last_modified": _mtime(f),
                }
                if hash_files:
                    book["file_hash"] = _hash(f)
                books.append(book)
        if not books:
            return (entry.as_uri(), None)
        return (entry.as_uri(), {
            "url": entry.as_uri(),
            "name": entry.name,
            "file_last_modified": _mtime(entry),
            "books": books,
        })
    except PermissionError as e:
        return (entry.as_uri(), None)


def walk_library(root: str, *, exclusions: set[str] | None = None,
                 oneshots_dir: str | None = None,
                 hash_files: bool = False,
                 max_workers: int | None = None,
                  on_progress: Callable | None = None) -> dict:
    """
    Walk a library root directory scanning for .pdf files only.

    Args:
        root: Filesystem path to the library root
        exclusions: Set of directory names to skip
        oneshots_dir: Subdirectory name for oneshots (e.g. "Oneshots")
        hash_files: If True, compute XXH3_128 hash for each PDF file
        max_workers: Number of parallel threads for directory scanning.
                     0 or None = auto (CPU count × 2). 1 = sequential.
        on_progress: Optional callback(completed, total) called after each
                     directory is processed.

    Returns:
    {
      "series": { "file:///root/SeriesA": { ... } },
      "oneshots": [...]
    }
    """
    if max_workers is None or max_workers == 0:
        max_workers = max((os.cpu_count() or 4) * 2, 1)

    root_path = Path(root).resolve()
    exclusions = exclusions or set()

    # Collect directories to scan
    dirs = [
        e for e in root_path.iterdir()
        if e.is_dir() and e.name not in exclusions
    ]

    # Parallel walk directories
    series: dict[str, dict] = {}
    if dirs:
        with ThreadPoolExecutor(max_workers=max_workers) as pool:
            futures = {
                pool.submit(_walk_one_dir, d, hash_files): d
                for d in dirs
            }
            completed = 0
            for future in as_completed(futures):
                completed += 1
                if on_progress:
                    on_progress(completed, len(dirs), futures[future].name)
                series_url, series_data = future.result()
                if series_data:
                    series[series_url] = series_data

    # Oneshots at root level (PDF only)
    oneshots: list[dict] = []
    for entry in sorted(root_path.iterdir(), key=lambda e: e.name.lower()):
        if entry.is_file() and entry.suffix.lower() == ".pdf":
            oneshot = {
                "url": entry.as_uri(),
                "name": entry.stem,
                "file_size": entry.stat().st_size,
                "file_last_modified": _mtime(entry),
            }
            if hash_files:
                oneshot["file_hash"] = _hash(entry)
            oneshots.append(oneshot)

    # Oneshots in dedicated oneshots directory
    if oneshots_dir:
        oneshots_path = root_path / oneshots_dir
        if oneshots_path.is_dir():
            for f in sorted(oneshots_path.iterdir(), key=lambda e: e.name.lower()):
                if f.is_file() and f.suffix.lower() == ".pdf":
                    oneshot = {
                        "url": f.as_uri(),
                        "name": f.stem,
                        "file_size": f.stat().st_size,
                        "file_last_modified": _mtime(f),
                    }
                    if hash_files:
                        oneshot["file_hash"] = _hash(f)
                    oneshots.append(oneshot)

    return {"series": series, "oneshots": oneshots}

File: scan-library-tool/rust/src/walker.rs
use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::sync::atomic::{AtomicUsize, Ordering};
use std::sync::Mutex;

use chrono::{DateTime, Utc};
use rayon::prelude::*;
use walkdir::WalkDir;

use crate::hasher;
use crate::models::{FsBook, FsData, FsSeries, HashCacheEntry};

fn mtime_to_iso(mtime: std::time::SystemTime) -> String {
    let dt: DateTime<Utc> = mtime.into();
    dt.to_rfc3339()
}

fn file_uri(path: &Path) -> String {
    url::Url::from_file_path(path)
        .map(|u| u.to_string())
        .unwrap_or_else(|_| format!("file://{}", path.display()))
}

fn compute_hash(
    path: &Path,
    hash_cache: &Option<HashMap<String, HashCacheEntry>>,
    hash_files: bool,
) -> Option<String> {
    if hash_files {
        if let Some(cache) = hash_cache {
            let uri = file_uri(path);
            if let Some(entry) = cache.get(&uri) {
                return Some(entry.hash.clone());
            }
        }
        // Fall back to computing the hash directly
        match hasher::hash_file(path, path.metadata().ok()?.len()) {
            Ok(h) => Some(h),
            Err(e) => {
                tracing::warn!(file = %path.display(), error = %e, "Hash computation failed");
                None
            }
        }
    } else {
        None
    }
}

fn scan_dir(
    dir: &Path,
    hash_cache: &Option<HashMap<String, HashCacheEntry>>,
    hash_files: bool,
) -> Option<FsSeries> {
    let meta = dir.metadata().ok()?;
    let mtime = meta.modified().ok()?;
    let file_last_modified = mtime_to_iso(mtime);

    let mut books = Vec::new();

    let entries: Vec<_> = WalkDir::new(dir)
        .min_depth(1)
        .follow_links(false)
        .into_iter()
        .filter_map(|e| e.ok())
        .filter(|e| {
            e.file_type().is_file()
                && e.path()
                    .extension()
                    .map_or(false, |ext| ext.eq_ignore_ascii_case("pdf"))
        })
        .collect();

    for entry in &entries {
        let fpath = entry.path();
        let Ok(fmeta) = fpath.metadata() else { continue };
        let Some(fmtime) = fmeta.modified().ok() else { continue };
        let url = file_uri(fpath);
        let name = fpath
            .file_stem()
            .map(|s| s.to_string_lossy().to_string())
            .unwrap_or_default();
        let file_hash = compute_hash(fpath, hash_cache, hash_files);

        books.push(FsBook {
            url,
            name,
            file_size: fmeta.len(),
            file_last_modified: mtime_to_iso(fmtime),
            file_hash,
        });
    }

    if books.is_empty() {
        return None;
    }

    // Sort books by name case-insensitively
    books.sort_by(|a, b| a.name.to_lowercase().cmp(&b.name.to_lowercase()));

    Some(FsSeries {
        url: file_uri(dir),
        name: dir
            .file_name()
            .map(|s| s.to_string_lossy().to_string())
            .unwrap_or_default(),
        file_last_modified,
        books,
    })
}

pub fn walk_library(
    root: &str,
    exclusions: &[String],
    oneshots_dir: Option<&str>,
    hash_files: bool,
    hash_cache: &Option<HashMap<String, HashCacheEntry>>,
    max_workers: usize,
    on_progress: Option<impl Fn(usize, usize, &str) + Send + Sync>,
) -> FsData {
    let root_path = PathBuf::from(root);
    let ex_set: std::collections::HashSet<&str> =
        exclusions.iter().map(|s| s.as_str()).collect();

    // Collect top-level directories
    let mut dirs: Vec<PathBuf> = Vec::new();
    if let Ok(read_dir) = root_path.read_dir() {
        for entry in read_dir.flatten() {
            let path = entry.path();
            if path.is_dir() {
                if let Some(name) = path.file_name().and_then(|n| n.to_str()) {
                    if !ex_set.contains(name) {
                        dirs.push(path);
                    }
                }
            }
        }
    }

    // Parallel walk directories
    let series_map = Mutex::new(HashMap::new());
    let completed = AtomicUsize::new(0);
    let total = dirs.len();

    if !dirs.is_empty() {
        let effective_threads = if max_workers == 0 {
            std::thread::available_parallelism()
                .map(|n| n.get() * 2)
                .unwrap_or(8)
        } else {
            max_workers
        };

        let pool = rayon::ThreadPoolBuilder::new()
            .num_threads(effective_threads)
            .build()
            .unwrap();

        pool.install(|| {
            dirs.par_iter().for_each(|dir| {
                let result = scan_dir(dir, hash_cache, hash_files);
                let idx = completed.fetch_add(1, Ordering::SeqCst) + 1;
                if let Some(ref cb) = on_progress {
                    let dir_name = dir
                        .file_name()
                        .map(|s| s.to_string_lossy().to_string())
                        .unwrap_or_default();
                    cb(idx, total, &dir_name);
                }
                if let Some(series) = result {
                    series_map.lock().unwrap().insert(series.url.clone(), series);
                }
            });
        });
    }

    // Oneshots at root level
    let mut oneshots: Vec<FsBook> = Vec::new();
    if let Ok(read_dir) = root_path.read_dir() {
        let mut entries: Vec<_> = read_dir
            .flatten()
            .filter(|e| {
                e.file_type().map(|t| t.is_file()).unwrap_or(false)
                    && e.path().extension().map_or(false, |ext| ext.eq_ignore_ascii_case("pdf"))
            })
            .collect();
        entries.sort_by_key(|e| e.file_name().to_ascii_lowercase());

        for entry in &entries {
            let fpath = entry.path();
            let Ok(fmeta) = fpath.metadata() else { continue };
            let Some(fmtime) = fmeta.modified().ok() else { continue };
            let url = file_uri(&fpath);
            let name = fpath
                .file_stem()
                .map(|s| s.to_string_lossy().to_string())
                .unwrap_or_default();
            let file_hash = compute_hash(&fpath, hash_cache, hash_files);

            oneshots.push(FsBook {
                url,
                name,
                file_size: fmeta.len(),
                file_last_modified: mtime_to_iso(fmtime),
                file_hash,
            });
        }
    }

    // Oneshots in dedicated oneshots directory
    if let Some(od) = oneshots_dir {
        let oneshots_path = root_path.join(od);
        if oneshots_path.is_dir() {
            if let Ok(read_dir) = oneshots_path.read_dir() {
                let mut entries: Vec<_> = read_dir
                    .flatten()
                    .filter(|e| {
                        e.file_type().map(|t| t.is_file()).unwrap_or(false)
                            && e.path().extension().map_or(false, |ext| ext.eq_ignore_ascii_case("pdf"))
                    })
                    .collect();
                entries.sort_by_key(|e| e.file_name().to_ascii_lowercase());

                for entry in &entries {
                    let fpath = entry.path();
                    let Ok(fmeta) = fpath.metadata() else { continue };
                    let Some(fmtime) = fmeta.modified().ok() else { continue };
                    let url = file_uri(&fpath);
                    let name = fpath
                        .file_stem()
                        .map(|s| s.to_string_lossy().to_string())
                        .unwrap_or_default();
                    let file_hash = compute_hash(&fpath, hash_cache, hash_files);

                    oneshots.push(FsBook {
                        url,
                        name,
                        file_size: fmeta.len(),
                        file_last_modified: mtime_to_iso(fmtime),
                        file_hash,
                    });
                }
            }
        }
    }

    FsData {
        series: series_map.into_inner().unwrap(),
        oneshots,
    }
}
